src/graph/sampling/randomwalks/node2vec_randomwalk.h

```diff
@@ -0,0 +1,218 @@
+    dgl_id_t,   // last last node id generated
+    int64_t)>;  // # of steps
+
+template <typename IdxType>
+bool has_edge_between(const std::vector<IdArray> &edges, dgl_id_t u,
+                      dgl_id_t v) {
+  const IdxType *offsets = static_cast<IdxType *>(edges[0]->data);
+  const IdxType *all_succ = static_cast<IdxType *>(edges[1]->data);
+  const IdxType *u_succ = all_succ + offsets[u];
+  const int64_t size = offsets[u + 1] - offsets[u];
+
+  return std::find(u_succ, u_succ + size, v) != u_succ + size;
+}
+
+/*!
+ * \brief Node2vec random walk step function
+ * \param data The path generated so far, of type \c IdxType.
+ * \param curr The last node ID generated.
+ * \param pre The last last node ID generated
+ * \param p Float, indicating likelihood of immediately revisiting a node in the
+ * walk. \param q Float, control parameter to interpolate between breadth-first
+ * strategy and depth-first strategy. \param len The number of nodes generated
+ * so far.  Note that the seed node is always included as \c data[0], and the
+ * successors start from \c data[1]. \param edges Vector of results from \c
+ * GetAdj() \param prob Transition probability \param terminate Predicate for
+ * terminating the current random walk path. \return A pair of ID of next
+ * successor (-1 if not exist), as well as whether to terminate.
+ */
+
+template <DLDeviceType XPU, typename IdxType>
+std::pair<dgl_id_t, bool> Node2vecRandomWalkStep(
+    IdxType *data, dgl_id_t curr, dgl_id_t pre, const double p, const double q,
+    int64_t len, const std::vector<IdArray> &edges, const FloatArray &probs,
+    TerminatePredicate<IdxType> terminate) {
+  const IdxType *offsets = static_cast<IdxType *>(edges[0]->data);
+  const IdxType *all_succ = static_cast<IdxType *>(edges[1]->data);
+  const IdxType *succ = all_succ + offsets[curr];
+
+  const int64_t size = offsets[curr + 1] - offsets[curr];
+
+  // Isolated node
+  if (size == 0) return std::make_pair(-1, true);
+
+  IdxType idx = 0;
+
+  // Normalize the weights to compute rejection probabilities
+  double max_prob = std::max({1 / p, 1.0, 1 / q});
+  // rejection prob for back to the previous node
+  double prob0 = 1 / p / max_prob;
+  // rejection prob for visiting the node with the distance of 1 between the
+  // previous node
+  double prob1 = 1 / max_prob;
+  // rejection prob for visiting the node with the distance of 2 between the
+  // previous node
+  double prob2 = 1 / q / max_prob;
+  dgl_id_t next_node;
+  double r;  // rejection probability.
+  if (IsNullArray(probs)) {
+    if (len == 0) {
+      idx = RandomEngine::ThreadLocal()->RandInt(size);
+      next_node = succ[idx];
+    } else {
+      while (true) {
+        idx = RandomEngine::ThreadLocal()->RandInt(size);
+        r = RandomEngine::ThreadLocal()->Uniform(0., 1.);
+        next_node = succ[idx];
+        if (next_node == pre) {
+          if (r < prob0) break;
+        } else if (has_edge_between<IdxType>(edges, next_node, pre)) {
```

Comment:
This is calling `has_edge_between` (i.e. `std::find`) repeatedly.  Did you compare with using `phmap::flat_hash_map` (a faster `unordered_map`)?